### ISSUE TITLE
EICrecon takes 20 minutes to build

### ISSUE THREAD

**Issue by veprbl** [eic/EICrecon#628]
```
time cmake --build build -- -j1
```
```output
real    22m48.523s
user    21m30.882s
sys     1m0.390s
```

**Comment by veprbl** [comment]
An output of compiling with `-DCMAKE_CXX_COMPILER_LAUNCHER=time -DCMAKE_CXX_LINKER_LAUNCHER=time` suggests that the time spent linking is negligible.

**Comment by veprbl** [comment]
Also, apparently, it's not related to datamodel_glue.h or datamodel_includes.h.

**Comment by nathanwbrei** [comment]
I've heard that clang has some nifty compile-time profiling tools, though I haven't tried them out yet. I'd be interested in how you profile this. Or maybe I should figure it out.

**Comment by veprbl** [comment]
> I've heard that clang has some nifty compile-time profiling tools, though I haven't tried them out yet.

Glad that you've posted this sentence. I wasn't able to figure out how to ask google about this. `-ftime-trace` looks promising.

**Comment by veprbl** [comment]
Looking at few files, Eigen is a big offender, but that's what it does. There are some "spdlog" and "JANA" headers that don't take much time (~1s?), but they should add up over the codebase. It came to me as a surprise, but a lot of time is spent in the "backend" (even for a `-DCMAKE_BUILD_TYPE=` build).

<img width="526" alt="image" src="https://user-images.githubusercontent.com/245573/234119559-c286d440-a2bc-4380-8c9b-6318841edb67.png">

**Comment by veprbl** [comment]
I've tried using include-what-you-use (IWYU), and it gave me about 1 minute speedup (5%). Not much, given that it requires quite a bit of tuning to get a nice and working diff.

**Comment by veprbl** [comment]
Sorry, not sure what happened. I didn't mean to forcefully assign anyone.

**Comment by nathanwbrei** [comment]
No worries! I just assumed you gave up in frustration and were hoping I'd think of something

**Comment by wdconinc** [comment]
Yeah, I noticed that GitHub had attached this to the #784 but I couldn't figure out how to detach it again. That PR explicitly doesn't address this and likely makes it worse :-D

**Comment by veprbl** [comment]
Robots don't get irony

**Comment by DraTeots** [comment]
Interesting insights using this tool: 
https://github.com/aras-p/ClangBuildAnalyzer

[eicrecon_compilation_analysis.txt](https://github.com/eic/EICrecon/files/12519551/eicrecon_compilation_analysis.txt)

**Comment by wdconinc** [comment]
Interesting...

```
**** Templates that took longest to instantiate:
134874 ms: nlohmann::basic_json<>::parse<const char *> (163 times, avg 827 ms)
 93952 ms: nlohmann::basic_json<>::basic_json (815 times, avg 115 ms)
 85261 ms: nlohmann::detail::parser<nlohmann::basic_json<>, nlohmann::detail::i... (163 times, avg 523 ms)
 71201 ms: nlohmann::detail::parser<nlohmann::basic_json<>, nlohmann::detail::i... (163 times, avg 436 ms)
 53404 ms: nlohmann::basic_json<>::json_value::json_value (817 times, avg 65 ms)
```
This is due to the `nlohmann/json.hpp` in every data model header, e.g. `/usr/local/include/edm4eic/TrackerHit.h` and `/usr/local/include/edm4hep/TrackerHit.h`. We should be using `nlohmann/json_fwd.hpp` at the very least for this exact reason, but even more surprising is that we have `PODIO_JSON_OUTPUT` set for all headers...

**Comment by wdconinc** [comment]
https://github.com/AIDASoft/podio/issues/475
